devito/types/array.py

```diff
@@ -345,3 +346,217 @@ def dim(self):
+        # size, hence we gotta use the actual numeric size instead
+        ret = []
+        for d, s, v in zip(self.dimensions, super().symbolic_shape, self.c0.shape):
+            if d.is_Stepping:
+                ret.append(Number(v))
+            else:
+                ret.append(s)
+        return DimensionTuple(*ret, getters=self.dimensions)
+
+    @property
+    def initvalue(self):
+        return None
+
+    @property
+    def is_shared(self):
+        if self.c0.is_Array:
+            return self.c0._is_shared
+        else:
+            return False
+
+    # CodeSymbol overrides defaulting to self.c0's behaviour
+
+    for i in ['_mem_internal_eager', '_mem_internal_lazy', '_mem_local',
+              '_mem_mapped', '_mem_host', '_mem_stack', '_size_domain',
+              '_size_halo', '_size_owned', '_size_padding', '_size_nopad',
+              '_size_nodomain', '_offset_domain', '_offset_halo', '_offset_owned',
+              '_dist_dimensions', '_C_get_field']:
+        locals()[i] = property(lambda self, v=i: getattr(self.c0, v))
+
+    @property
+    def _mem_heap(self):
+        return not self._mem_stack
+
+    @property
+    def _dist_dimensions(self):
+        return self.c0._dist_dimensions
+
+    def _C_get_field(self, region, dim, side=None):
+        return self.c0._C_get_field(region, dim, side=side)
+
+    def __getitem__(self, index):
+        index = as_tuple(index)
+        if len(index) == self.ndim:
+            return super().__getitem__(index)
+        elif len(index) == self.ndim + 1:
+            component_index, indices = index[0], index[1:]
+            return ComponentAccess(self.indexed[indices], component_index)
+        else:
+            raise ValueError("Expected %d or %d indices, got %d instead"
+                             % (self.ndim, self.ndim + 1, len(index)))
+
+    _C_structname = ArrayMapped._C_structname
+    _C_field_data = ArrayMapped._C_field_data
+    _C_field_nbytes = ArrayMapped._C_field_nbytes
+    _C_field_dmap = ArrayMapped._C_field_dmap
+
+    @property
+    def _C_ctype(self):
+        if self._mem_mapped:
+            return ArrayMapped._C_ctype
+        else:
+            return POINTER(dtype_to_ctype(self.dtype))
+
+
+class ComponentAccess(Expr, Reconstructable):
+
+    _component_names = ('x', 'y', 'z', 'w')
+
+    __rkwargs__ = ('index',)
+
+    def __new__(cls, arg, index=0, **kwargs):
+        if not arg.is_Indexed:
+            raise ValueError("Expected Indexed, got `%s` instead" % type(arg))
+        if not isinstance(index, int) or index > 3:
+            raise ValueError("Expected 0 <= index < 4")
+
+        obj = Expr.__new__(cls, arg)
+        obj._index = index
+
+        return obj
+
+    def _hashable_content(self):
+        return super()._hashable_content() + (self._index,)
+
+    def __str__(self):
+        return "%s.%s" % (self.base, self.sindex)
+
+    __repr__ = __str__
+
+    func = Reconstructable._rebuild
+
+    def _sympystr(self, printer):
```

Comment:
Still not sure why need this, everything in sympy goes through `_print` which should be safer to use no?

Reply:
see comment above, `_print` is only when you use `str(..)` basically